gson/src/main/java/com/google/gson/internal/bind/DefaultDateTypeAdapter.java

```diff
@@ -134,6 +134,8 @@ private DefaultDateTypeAdapter(DateType<T> dateType, int dateStyle, int timeStyl
     if (JavaVersion.isJava9OrLater()) {
       dateFormats.add(PreJava9DateFormatProvider.getUsDateTimeFormat(dateStyle, timeStyle));
     }
+    // Include fall-back for "timeless" Date (could be from a java.sql.Date, for example)
+    dateFormats.add(DateFormat.getDateInstance(dateStyle, Locale.US));
```

Comment:
This seems a bit brittle for the use case you are describing because the `java.sql.Date` adapter actually uses a hardcoded pattern (with default locale) and not a date style:
https://github.com/google/gson/blob/18b08922537c16b0de1d503c82307d613aae0138/gson/src/main/java/com/google/gson/internal/sql/SqlDateTypeAdapter.java#L52

Reply:
> I have to basically copy the exact `DefaultDateTypeAdapter` and make this one change

You can avoid this by writing a [`TypeAdapterFactory`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/TypeAdapterFactory.html) which first tries to deserialize as `java.util.Date` and if that fails tries `java.sql.Date` instead:
<details>
<summary><code>DateSqlFallbackFactory</code> (click to expand)</summary>

Note that I haven't tested this extensively, and this implementation is also quite verbose. And if parsing fails, the JSON path in the exception message will not be that helpful, saying `$` (root element) instead of the actual path.

```java
class DateSqlFallbackFactory implements TypeAdapterFactory {
  @Override
  public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
    if (type.getRawType() != java.util.Date.class) {
      return null;
    }

    TypeAdapter<JsonElement> jsonElementAdapter = gson.getAdapter(JsonElement.class);
    TypeAdapter<Date> utilDateAdapter = gson.getDelegateAdapter(this, TypeToken.get(Date.class));
    TypeAdapter<java.sql.Date> sqlDateAdapter = gson.getAdapter(java.sql.Date.class);

    @SuppressWarnings("unchecked")
    TypeAdapter<T> adapter = (TypeAdapter<T>) new TypeAdapter<Date>() {
      @Override
      public Date read(JsonReader in) throws IOException {
        // First read as JsonElement tree to be able to parse it twice (once as java.util.Date, once as
        // java.sql.Date) if necessary
        JsonElement json = jsonElementAdapter.read(in);
        try {
          return utilDateAdapter.fromJsonTree(json);
        }
        // Note: This makes assumptions about the exception type thrown by Gson's built-in
        // adapter for java.util.Date
        catch (JsonParseException e) {
          try {
            return sqlDateAdapter.fromJsonTree(json);
          } catch (Exception suppressed) {
            e.addSuppressed(suppressed);
            throw e;
          }
        }
      }

      @Override
      public void write(JsonWriter out, Date value) throws IOException {
        utilDateAdapter.write(out, value);
      }
    };
    return adapter;
  }
}
```

</details>

----

If you haven't stored the JSON data for any `java.util.Date` and SQL subclasses persistently yet, then maybe a reliable alternative would be to use [`GsonBuilder.setDateFormat(String)`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/GsonBuilder.html#setDateFormat(java.lang.String)) to specify a locale insensitive pattern (if possible).